File: pipeline/run_pipeline.py
import os
import json
import argparse
from datetime import datetime

from surgeon.simulate_calls import simulate_call, save_simulation
from detective.evaluator import evaluate_call


def load_prompt(prompt_path):
    with open(prompt_path, "r", encoding="utf-8") as f:
        return f.read()


def load_transcripts(folder):

    calls = []

    # Files to skip — not real transcripts
    SKIP_FILES = {"_manifest.json", "manifest.json"}

    for file in os.listdir(folder):

        if file.endswith(".json") and file not in SKIP_FILES:

            with open(os.path.join(folder, file), "r", encoding="utf-8") as f:
                calls.append(json.load(f))

    return calls


def run_pipeline(prompt_path, transcript_folder):

    prompt = load_prompt(prompt_path)
    transcripts = load_transcripts(transcript_folder)

    results = []

    for call in transcripts:

        call_id = call["call_id"]

        print(f"\nRunning {call_id}...")

        simulated = simulate_call(prompt, call)

        sim_file = save_simulation(call_id, simulated)

        score, issues, verdict = evaluate_call(simulated)

        results.append({
            "call_id": call_id,
            "score": score,
            "verdict": verdict,
            "issues": issues,
            "simulation_file": sim_file
        })

        print(f"Score: {score} | Verdict: {verdict}")

    return results


def generate_report(results):

    os.makedirs("results", exist_ok=True)

    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")

    avg_score = sum(r["score"] for r in results) / len(results)

    good = sum(1 for r in results if r["verdict"] == "good")
    bad = sum(1 for r in results if r["verdict"] == "bad")

    report = {
        "average_score": avg_score,
        "good_calls": good,
        "bad_calls": bad,
        "results": results
    }

    report_file = f"results/report_{timestamp}.json"

    with open(report_file, "w", encoding="utf-8") as f:
        json.dump(report, f, indent=2, ensure_ascii=False)

    print("\nPipeline Report")
    print("----------------")
    print("Average Score:", avg_score)
    print("Good Calls:", good)
    print("Bad Calls:", bad)
    print("Saved:", report_file)


if __name__ == "__main__":

    parser = argparse.ArgumentParser()

    parser.add_argument("--prompt", required=True)
    parser.add_argument("--transcripts", required=True)

    args = parser.parse_args()

    results = run_pipeline(args.prompt, args.transcripts)

    generate_report(results)

File: pipeline/run_pipeline.js
import fs from "fs";
import path from "path";
import { simulateCall, saveSimulation } from "../surgeon/simulate_calls.js";
import { evaluateCall } from "../detective/evaluator.js";

// ---------------------------------------------------------------------------
// Load Prompt
// ---------------------------------------------------------------------------

function loadPrompt(promptPath) {
  return fs.readFileSync(promptPath, "utf-8");
}

// ---------------------------------------------------------------------------
// Load Transcripts
// ---------------------------------------------------------------------------

function loadTranscripts(folder) {
  const calls = [];

  const SKIP_FILES = new Set(["_manifest.json", "manifest.json"]);

  const files = fs.readdirSync(folder);

  for (const file of files) {
    if (file.endsWith(".json") && !SKIP_FILES.has(file)) {
      const content = fs.readFileSync(path.join(folder, file), "utf-8");
      calls.push(JSON.parse(content));
    }
  }

  return calls;
}

// ---------------------------------------------------------------------------
// Run Pipeline
// ---------------------------------------------------------------------------

export async function runPipeline(promptPath, transcriptFolder) {
  const prompt = loadPrompt(promptPath);
  const transcripts = loadTranscripts(transcriptFolder);

  const results = [];

  for (const call of transcripts) {
    const callId = call.call_id;

    console.log(`\nRunning ${callId}...`);

    const simulated = await simulateCall(prompt, call);

    const simFile = await saveSimulation(callId, simulated);

    const { score, issues, verdict } = await evaluateCall(simulated);

    results.push({
      call_id: callId,
      score,
      verdict,
      issues,
      simulation_file: simFile,
    });

    console.log(`Score: ${score} | Verdict: ${verdict}`);
  }

  return results;
}

// ---------------------------------------------------------------------------
// Generate Report
// ---------------------------------------------------------------------------

export function generateReport(results) {
  fs.mkdirSync("results", { recursive: true });

  const timestamp = new Date()
    .toISOString()
    .replace(/[-:]/g, "")
    .slice(0, 15);

  const avgScore =
    results.reduce((sum, r) => sum + r.score, 0) / results.length;

  const good = results.filter((r) => r.verdict === "good").length;
  const bad = results.filter((r) => r.verdict === "bad").length;

  const report = {
    average_score: avgScore,
    good_calls: good,
    bad_calls: bad,
    results,
  };

  const reportFile = `results/report_${timestamp}.json`;

  fs.writeFileSync(reportFile, JSON.stringify(report, null, 2));

  console.log("\nPipeline Report");
  console.log("----------------");
  console.log("Average Score:", avgScore);
  console.log("Good Calls:", good);
  console.log("Bad Calls:", bad);
  console.log("Saved:", reportFile);
}

// ---------------------------------------------------------------------------
// CLI
// ---------------------------------------------------------------------------

async function main() {
  const args = process.argv.slice(2);

  const promptIndex = args.indexOf("--prompt");
  const transcriptsIndex = args.indexOf("--transcripts");

  if (promptIndex === -1 || transcriptsIndex === -1) {
    console.error(
      "Usage: node runPipeline.js --prompt <prompt_file> --transcripts <folder>"
    );
    process.exit(1);
  }

  const promptPath = args[promptIndex + 1];
  const transcriptsFolder = args[transcriptsIndex + 1];

  const results = await runPipeline(promptPath, transcriptsFolder);

  generateReport(results);
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main();
}